File: python/src/threads/adapters/sandboxes/dev/walk.py
"""Every host-side file operation of the dev sandbox walks the path with `openat`, one component
at a time, each opened `O_NOFOLLOW`, so no symlink is ever traversed. A lexical check can't do
this: `a/b -> ..` followed by `c -> a/b/..` leaves the root through two links that each look
harmless. A symlink is only ever read as a link here, never traversed. Inside the confinement the
command may follow links freely: that is the sandbox's own filesystem.
"""

import errno
import os
import posixpath
import stat
from collections.abc import Generator, Sequence
from contextlib import contextmanager
from dataclasses import dataclass
from typing import Literal

from threads.result import Err, Ok
from threads.sandbox.protocol import SandboxError, SandboxErrorCode

WORKSPACE = "/workspace"

_DIR_FLAGS = os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW

_CODES: dict[int, SandboxErrorCode] = {
    errno.ENOENT: "not_found",
    errno.ENOTDIR: "not_found",
    errno.EISDIR: "is_directory",
    errno.ELOOP: "invalid_path",
    errno.EMLINK: "invalid_path",
    errno.EACCES: "permission_denied",
    errno.EPERM: "permission_denied",
    errno.EFBIG: "too_large",
}


def failed(error: OSError, path: str) -> SandboxError:
    """A host errno as the protocol's file failure."""
    code = _CODES.get(error.errno or 0, "unavailable")
    return SandboxError(code, f"{path}: {errno.errorcode.get(error.errno or 0, error.strerror)}")


def _linked(path: str) -> SandboxError:
    return SandboxError("invalid_path", f"{path}: a symlink is never followed on the host")


def workspace_parts(path: str) -> Ok[tuple[str, ...]] | Err[SandboxError]:
    """The components of `path` under /workspace, or a typed failure: a path outside /workspace is
    outside the sandbox, because the dev sandbox's filesystem is the host's."""
    if "\0" in path or not path.startswith("/"):
        return Err(SandboxError("invalid_path", f"not an absolute path: {path}"))
    at = posixpath.normpath(path)
    if at != WORKSPACE and not at.startswith(f"{WORKSPACE}/"):
        return Err(SandboxError("invalid_path", f"{path}: the dev sandbox holds only {WORKSPACE}"))
    return Ok(tuple(part for part in at[len(WORKSPACE) :].split("/") if part))


def _is_link(fd: int, name: str) -> bool:
    """Whether `name` in `fd` is a symlink. O_NOFOLLOW's errno differs by platform (ELOOP on
    Linux, ENOTDIR with O_DIRECTORY on macOS), so the refusal is named from the link itself."""
    try:
        return stat.S_ISLNK(os.lstat(name, dir_fd=fd).st_mode)
    except OSError:
        return False


def _child_dir(fd: int, name: str, at: str, create: bool) -> Ok[int] | Err[SandboxError]:
    """An fd on `name` inside `fd`, refusing a symlink; makes it when `create`."""
    try:
        return Ok(os.open(name, _DIR_FLAGS, dir_fd=fd))
    except OSError as error:
        if _is_link(fd, name):
            return Err(_linked(at))
        if error.errno == errno.ENOENT and create:
            try:
                os.mkdir(name, 0o755, dir_fd=fd)
                return Ok(os.open(name, _DIR_FLAGS, dir_fd=fd))
            except OSError as made:
                return Err(failed(made, at))
        return Err(failed(error, at))


@dataclass(frozen=True, slots=True)
class Walked:
    """The directory holding the last component, and that component's name."""

    fd: int
    name: str


@contextmanager
def walked(
    directory: str, parts: Sequence[str], create: bool
) -> Generator[Ok[Walked] | Err[SandboxError]]:
    """`directory` walked down to the parent of `parts[-1]`, every component opened
    `O_NOFOLLOW`. The fd is closed when the block ends."""
    try:
        fd = os.open(directory, _DIR_FLAGS)
    except OSError as error:
        yield Err(failed(error, directory))
        return
    at = directory
    try:
        for part in parts[:-1]:
            at = f"{at}/{part}"
            child = _child_dir(fd, part, at, create)
            if isinstance(child, Err):
                yield child
                return
            os.close(fd)
            fd = child.value
        yield Ok(Walked(fd, parts[-1] if parts else "."))
    finally:
        os.close(fd)


def host_dir(directory: str, parts: Sequence[str]) -> Ok[str] | Err[SandboxError]:
    """The host path of a directory under `directory`, every component of it proven to be a real
    directory. An exec's cwd goes through this, so no symlink chain ever chooses where a command
    starts."""
    with walked(directory, (*parts, "."), create=False) as found:
        if isinstance(found, Err):
            return found
        return Ok(os.path.join(directory, *parts))


def read_in(directory: str, parts: Sequence[str]) -> Ok[bytes] | Err[SandboxError]:
    """The file's bytes, refusing a symlink at any component."""
    with walked(directory, parts, create=False) as found:
        if isinstance(found, Err):
            return found
        at = "/".join((directory, *parts))
        try:
            fd = os.open(found.value.name, os.O_RDONLY | os.O_NOFOLLOW, dir_fd=found.value.fd)
        except OSError as error:
            if _is_link(found.value.fd, found.value.name):
                return Err(_linked(at))
            return Err(failed(error, at))
        try:
            if stat.S_ISDIR(os.fstat(fd).st_mode):
                return Err(SandboxError("is_directory", at))
            return Ok(_read_all(fd))
        except OSError as error:
            return Err(failed(error, at))
        finally:
            os.close(fd)


def _read_all(fd: int) -> bytes:
    parts: list[bytes] = []
    while chunk := os.read(fd, 1 << 20):
        parts.append(chunk)
    return b"".join(parts)


def write_in(
    directory: str, parts: Sequence[str], data: bytes, mode: int
) -> Ok[None] | Err[SandboxError]:
    """Writes the file, making missing parents, refusing a symlink at any component."""
    with walked(directory, parts, create=True) as found:
        if isinstance(found, Err):
            return found
        at = "/".join((directory, *parts))
        flags = os.O_WRONLY | os.O_CREAT | os.O_TRUNC | os.O_NOFOLLOW
        try:
            fd = os.open(found.value.name, flags, mode, dir_fd=found.value.fd)
        except OSError as error:
            if _is_link(found.value.fd, found.value.name):
                return Err(_linked(at))
            return Err(failed(error, at))
        try:
            os.write(fd, data)
            return Ok(None)
        except OSError as error:
            return Err(failed(error, at))
        finally:
            os.close(fd)


def mkdir_in(directory: str, parts: Sequence[str], mode: int) -> Ok[None] | Err[SandboxError]:
    """Makes the directory, making missing parents, refusing a symlink at any component."""
    with walked(directory, parts, create=True) as found:
        if isinstance(found, Err):
            return found
        try:
            os.mkdir(found.value.name, mode, dir_fd=found.value.fd)
        except FileExistsError:
            return Ok(None)
        except OSError as error:
            return Err(failed(error, "/".join((directory, *parts))))
        return Ok(None)


def symlink_in(directory: str, parts: Sequence[str], target: str) -> Ok[None] | Err[SandboxError]:
    """Makes the symlink, refusing a symlink at any component of its own path."""
    with walked(directory, parts, create=True) as found:
        if isinstance(found, Err):
            return found
        try:
            os.symlink(target, found.value.name, dir_fd=found.value.fd)
        except OSError as error:
            return Err(failed(error, "/".join((directory, *parts))))
        return Ok(None)


@dataclass(frozen=True, slots=True)
class Found:
    """One thing in the tree, by its path relative to the sandbox's /workspace."""

    path: str
    kind: Literal["dir", "file", "symlink"]
    mode: int = 0
    size: int = 0
    target: str = ""


def walk_tree(directory: str) -> Ok[tuple[Found, ...]] | Err[SandboxError]:
    """Everything under `directory`, sorted by path, never descending into a symlink: each level
    is listed through its own `O_NOFOLLOW` fd."""
    out: list[Found] = []
    queue: list[tuple[str, ...]] = [()]
    while queue:
        prefix = queue.pop(0)
        listed = _level(directory, prefix, out)
        if isinstance(listed, Err):
            return listed
        queue.extend((*prefix, name) for name in listed.value)
    return Ok(tuple(sorted(out, key=lambda e: e.path)))


def _level(
    directory: str, prefix: Sequence[str], out: list[Found]
) -> Ok[tuple[str, ...]] | Err[SandboxError]:
    """One directory's entries appended to `out`; the names of its subdirectories."""
    with walked(directory, (*prefix, "."), create=False) as found:
        if isinstance(found, Err):
            return found
        fd = found.value.fd
        at = "/".join((directory, *prefix))
        try:
            names = sorted(os.listdir(fd))
        except OSError as error:
            return Err(failed(error, at))
        dirs: list[str] = []
        for name in names:
            entry = _entry(fd, prefix, name, at)
            if isinstance(entry, Err):
                return entry
            out.append(entry.value)
            if entry.value.kind == "dir":
                dirs.append(name)
        return Ok(tuple(dirs))


def _entry(fd: int, prefix: Sequence[str], name: str, at: str) -> Ok[Found] | Err[SandboxError]:
    path = "/".join((*prefix, name))
    try:
        info = os.lstat(name, dir_fd=fd)
        if stat.S_ISLNK(info.st_mode):
            return Ok(Found(path, "symlink", target=os.readlink(name, dir_fd=fd)))
    except OSError as error:
        return Err(failed(error, f"{at}/{name}"))
    mode = stat.S_IMODE(info.st_mode)
    if stat.S_ISDIR(info.st_mode):
        return Ok(Found(path, "dir", mode=mode))
    if stat.S_ISREG(info.st_mode):
        return Ok(Found(path, "file", mode=mode, size=info.st_size))
    return Err(SandboxError("invalid_path", f"{at}/{name}: not a file, directory or symlink"))

File: typescript/packages/core/src/sandbox/dev/walk.ts
import {
  closeSync,
  constants,
  type Dirent,
  lstatSync,
  mkdirSync,
  openSync,
  readdirSync,
  readFileSync,
  readlinkSync,
  symlinkSync,
  writeFileSync,
} from "node:fs";
import { err, ok, type Result } from "../../result";
import type { FileFailure } from "../protocol";
import { sandboxPath, WORKSPACE } from "../remote/scripts";

// Every host-side file operation of the dev sandbox walks the path one component at a time and
// refuses a component that is a symlink, then opens the last one with O_NOFOLLOW. A lexical
// check can't do this: `a/b -> ..` followed by `c -> a/b/..` leaves the root through two links
// that each look harmless. A symlink is only ever read as a link here, never traversed. Inside
// the confinement the command may follow links freely: that is the sandbox's own filesystem.
//
// ponytail: Node has no openat, so a component is checked with lstat and the last is opened
// with O_NOFOLLOW; a component swapped between the check and the open is still possible.
// Python's side does the real openat walk (adapters/sandboxes/dev/walk.py).

const NOFOLLOW = constants.O_NOFOLLOW;

const CODES: ReadonlyMap<string, FileFailure["code"]> = new Map([
  ["ENOENT", "not_found"],
  ["ENOTDIR", "not_found"],
  ["EISDIR", "is_directory"],
  ["ELOOP", "invalid_path"],
  ["EMLINK", "invalid_path"],
  ["EACCES", "permission_denied"],
  ["EPERM", "permission_denied"],
  ["EFBIG", "too_large"],
]);

function codeOf(error: unknown): string {
  if (error === null || typeof error !== "object" || !("code" in error))
    return "";
  return typeof error.code === "string" ? error.code : "";
}

/** A host errno as the protocol's file failure. */
export function failed(error: unknown, path: string): FileFailure {
  const errno = codeOf(error);
  return {
    code: CODES.get(errno) ?? "unavailable",
    message: `${path}: ${errno === "" ? String(error) : errno}`,
  };
}

const linked = (path: string): FileFailure => ({
  code: "invalid_path",
  message: `${path}: a symlink is never followed on the host`,
});

/**
 * The components of `path` under /workspace, or a typed failure: a path outside /workspace is
 * outside the sandbox, because the dev sandbox's filesystem is the host's.
 */
export function workspaceParts(
  path: string,
): Result<readonly string[], FileFailure> {
  const at = sandboxPath(path);
  if (at === undefined || (at !== WORKSPACE && !at.startsWith(`${WORKSPACE}/`)))
    return err({
      code: "invalid_path",
      message: `${path}: the dev sandbox holds only ${WORKSPACE}`,
    });
  return ok(
    at
      .slice(WORKSPACE.length)
      .split("/")
      .filter((part) => part !== ""),
  );
}

/**
 * `dir` joined with `parts`, every component before the last proven to be a real directory.
 * `create` makes a missing one (0o755), as the kits' write check does.
 */
export function hostPath(
  dir: string,
  parts: readonly string[],
  create: boolean,
): Result<string, FileFailure> {
  let at = dir;
  for (const part of parts.slice(0, -1)) {
    at = `${at}/${part}`;
    let stat: ReturnType<typeof lstatSync>;
    try {
      stat = lstatSync(at);
    } catch (error) {
      if (!create || codeOf(error) !== "ENOENT") return err(failed(error, at));
      try {
        mkdirSync(at, 0o755);
      } catch (made) {
        return err(failed(made, at));
      }
      continue;
    }
    if (stat.isSymbolicLink()) return err(linked(at));
    if (!stat.isDirectory())
      return err({ code: "not_found", message: `${at}: not a directory` });
  }
  const last = parts.at(-1);
  return ok(last === undefined ? at : `${at}/${last}`);
}

/** The file's bytes, refusing a symlink at any component. */
export function readIn(
  dir: string,
  parts: readonly string[],
): Result<Uint8Array, FileFailure> {
  const at = hostPath(dir, parts, false);
  if (!at.ok) return at;
  let fd: number | undefined;
  try {
    fd = openSync(at.value, constants.O_RDONLY | NOFOLLOW);
    return ok(new Uint8Array(readFileSync(fd)));
  } catch (error) {
    return err(failed(error, at.value));
  } finally {
    if (fd !== undefined) closeSync(fd);
  }
}

/** Writes the file, making missing parents, refusing a symlink at any component. */
export function writeIn(
  dir: string,
  parts: readonly string[],
  data: Uint8Array,
  mode: number,
): Result<void, FileFailure> {
  const at = hostPath(dir, parts, true);
  if (!at.ok) return at;
  let fd: number | undefined;
  try {
    const flags =
      constants.O_WRONLY | constants.O_CREAT | constants.O_TRUNC | NOFOLLOW;
    fd = openSync(at.value, flags, mode);
    writeFileSync(fd, data);
    return ok(undefined);
  } catch (error) {
    return err(failed(error, at.value));
  } finally {
    if (fd !== undefined) closeSync(fd);
  }
}

/** Makes the directory, making missing parents, refusing a symlink at any component. */
export function mkdirIn(
  dir: string,
  parts: readonly string[],
  mode: number,
): Result<void, FileFailure> {
  const at = hostPath(dir, parts, true);
  if (!at.ok) return at;
  try {
    mkdirSync(at.value, mode);
    return ok(undefined);
  } catch (error) {
    if (codeOf(error) === "EEXIST") return ok(undefined);
    return err(failed(error, at.value));
  }
}

/** Makes the symlink, refusing a symlink at any component of its own path. */
export function symlinkIn(
  dir: string,
  parts: readonly string[],
  target: string,
): Result<void, FileFailure> {
  const at = hostPath(dir, parts, true);
  if (!at.ok) return at;
  try {
    symlinkSync(target, at.value);
    return ok(undefined);
  } catch (error) {
    return err(failed(error, at.value));
  }
}

/** One thing found in the tree, by its path relative to the sandbox's /workspace. */
export type Found =
  | { readonly path: string; readonly kind: "dir"; readonly mode: number }
  | {
      readonly path: string;
      readonly kind: "file";
      readonly mode: number;
      readonly size: number;
      readonly host: string;
    }
  | {
      readonly path: string;
      readonly kind: "symlink";
      readonly target: string;
    };

function entryOf(dir: string, at: string, e: Dirent): Found | FileFailure {
  const host = `${dir}/${at}`;
  if (e.isSymbolicLink())
    return { path: at, kind: "symlink", target: readlinkSync(host) };
  const stat = lstatSync(host);
  const mode = stat.mode & 0o7777;
  if (e.isDirectory()) return { path: at, kind: "dir", mode };
  if (e.isFile())
    return { path: at, kind: "file", mode, size: stat.size, host };
  return {
    code: "invalid_path",
    message: `${host}: not a file, directory or symlink`,
  };
}

/**
 * Everything under `dir`, breadth by directory, sorted by path, never descending into a
 * symlink. Each entry's host path is read back through this same walk, so nothing outside the
 * root is ever opened.
 */
export function walkTree(dir: string): Result<readonly Found[], FileFailure> {
  const out: Found[] = [];
  const queue: string[] = [""];
  try {
    for (const prefix of queue) {
      const here = prefix === "" ? dir : `${dir}/${prefix}`;
      for (const e of readdirSync(here, { withFileTypes: true }).toSorted(
        (a, b) => (a.name < b.name ? -1 : 1),
      )) {
        const at = prefix === "" ? e.name : `${prefix}/${e.name}`;
        const found = entryOf(dir, at, e);
        if ("code" in found) return err(found);
        out.push(found);
        if (found.kind === "dir") queue.push(at);
      }
    }
  } catch (error) {
    return err(failed(error, dir));
  }
  return ok(out.toSorted((a, b) => (a.path < b.path ? -1 : 1)));
}
